examples/reference/custom_components/JSComponent.md

```diff
@@ -0,0 +1,425 @@
+# `JSComponent`
+
+`JSComponent` simplifies the creation of custom Panel components using JavaScript.
+
+```pyodide
+import panel as pn
+import param
+
+from panel.custom import JSComponent
+
+pn.extension()
+
+class CounterButton(JSComponent):
+
+    value = param.Integer()
+
+    _esm = """
+    export function render({ data }) {
+        let btn = document.createElement("button");
+        btn.innerHTML = `count is ${data.value}`;
+        btn.addEventListener("click", () => {
+            data.value += 1
+        });
+        data.watch(() => {
+            btn.innerHTML = `count is ${data.value}`;
+          }, 'value')
+        return btn
+    }
+    """
+
+CounterButton().servable()
+```
+
+:::{note}
+
+`JSComponent` was introduced in June 2024 as the successor to `ReactiveHTML`.
+
+`JSComponent` bears similarities to [`AnyWidget`](https://anywidget.dev/), but it is specifically optimized for use with Panel.
+
+:::
+
+## API
+
+### JSComponent Attributes
+
+- **`_esm`** (str | PurePath): This attribute accepts either a string or a path that points to an  [ECMAScript module](https://nodejs.org/api/esm.html#modules-ecmascript-modules). The ECMAScript module should export a `render` function which returns the HTML element to display. In a development environment such as a notebook or when using `--autoreload`, the module will automatically reload upon saving changes.
+- **`_import_map`** (dict | None): This optional dictionary defines an [import map](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap), allowing you to customize how module specifiers are resolved.
+- **`_stylesheets`** (optional list of strings): This optional attribute accepts a list of CSS strings or paths to CSS files. It supports automatic reloading in development environments.
+
+#### `render` Function
+
+The `_esm` attribute must export the `render` function. It accepts the following parameters:
```

Comment:
Please note that one day someone (like chat gpt right now) would like to use `async` methods in the `render` function.

Do you believe we can and should support this now or later?

Reply:
Should be possible, just have to check if the returned value is a Promise.